src/adapter/src/rbac.rs

```diff
@@ -771,85 +839,113 @@ fn generate_read_privileges(
     catalog: &impl SessionCatalog,
     ids: impl Iterator<Item = GlobalId>,
     role_id: RoleId,
-) -> BTreeSet<(ObjectId, AclMode, RoleId)> {
+) -> Vec<(ObjectId, AclMode, RoleId)> {
     generate_read_privileges_inner(catalog, ids, role_id, &mut BTreeSet::new())
 }
 
 fn generate_read_privileges_inner(
     catalog: &impl SessionCatalog,
     ids: impl Iterator<Item = GlobalId>,
     role_id: RoleId,
-    seen: &mut BTreeSet<(GlobalId, RoleId)>,
-) -> BTreeSet<(ObjectId, AclMode, RoleId)> {
-    let mut privileges = BTreeSet::new();
+    seen: &mut BTreeSet<(ObjectId, RoleId)>,
+) -> Vec<(ObjectId, AclMode, RoleId)> {
+    let mut privileges = Vec::new();
+    let mut views = Vec::new();
+
     for id in ids {
-        if seen.insert((id, role_id)) {
-            privileges
-                .extend(generate_read_privilege_inner(catalog, id, role_id, seen).into_iter());
+        if seen.insert((id.into(), role_id)) {
+            let item = catalog.get_item(&id);
+            let schema_id: ObjectId = item.name().qualifiers.clone().into();
+            if seen.insert((schema_id.clone(), role_id)) {
+                privileges.push((schema_id, AclMode::USAGE, role_id))
+            }
+            match item.item_type() {
+                CatalogItemType::View | CatalogItemType::MaterializedView => {
+                    privileges.push((id.into(), AclMode::SELECT, role_id));
+                    views.push((item.uses().iter().cloned(), item.owner_id()));
+                }
+                CatalogItemType::Table | CatalogItemType::Source => {
+                    privileges.push((id.into(), AclMode::SELECT, role_id));
+                }
+                CatalogItemType::Type | CatalogItemType::Secret | CatalogItemType::Connection => {
+                    privileges.push((id.into(), AclMode::USAGE, role_id));
+                }
+                CatalogItemType::Sink | CatalogItemType::Index | CatalogItemType::Func => {}
```

Comment:
Intentionally unimplemented?

Reply:
Yes, these objects do not have any associated privileges.